### PR TITLE
Add StatefulOutputPin and ToggleableOutputPin

### PR DESCRIPTION
**Pull request by Sh3Rm4n**
I tested it on the STM32F303 / STM32DISCOVERY Board, with the method provided in the example.

### CODE REVIEW

**Comment by Sh3Rm4n** [comment]
NOTE: I shamelessly stole the implementation from https://github.com/stm32-rs/stm32f1xx-hal. 
As this is, as of my understanding, a generic method to read out the state of the pin, this should work for every stm32 gpio register.

**Comment by dfrankland** [comment]
This looks great. I'll be looking to get this in once we test out and merge #21 since it has some hefty changes to `gpio.rs`.

File: examples/toggle.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+use stm32f3xx_hal::stm32;
+
+#[entry]
+fn main() -> ! {
+    let p = stm32::Peripherals::take().unwrap();
+
+    let mut rcc = p.RCC.constrain();
+    let mut gpioe = p.GPIOE.split(&mut rcc.ahb);
+
+    let mut led_red = gpioe
```
Comment (dfrankland) [inline examples/toggle.rs]: Might we be able to name this just `led`? The LED on my board is white :stuck_out_tongue_closed_eyes:

File: examples/toggle.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+//! Uses the StatefulOutputPin embedded_hal trait to toggle the pin
+//! On the stm32 discovery board this is the red led
+//! Target board: STM32F3DISCOVERY
+
+#![deny(unsafe_code)]
+#![no_main]
+#![no_std]
+
+use panic_semihosting as _;
```
Comment (dfrankland) [inline examples/toggle.rs]: To stay consistent with:
https://github.com/stm32-rs/stm32f3xx-hal/blob/902b56f8f11221cb8ffbf24305d66ecea858bf6c/examples/usb_serial.rs#L6

```suggestion
extern crate panic_semihosting;
```

File: src/prelude.rs [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 pub use crate::flash::FlashExt as _stm32f3xx_hal_flash_FlashExt;
 pub use crate::gpio::GpioExt as _stm32f3xx_hal_gpio_GpioExt;
+pub use crate::hal::digital::v2::StatefulOutputPin as _embedded_hal_digital_StatefulOutputPin;
+pub use crate::hal::digital::v2::ToggleableOutputPin as _embedded_hal_digital_ToggleableOutputPin;
```
Comment (dfrankland) [inline src/prelude.rs]: These re-exports needed to be guarded by `#[cfg(feature = "unproven")]`, right? Might be nice to re-export the `OutputPin` and `InputPin` traits too.

File: examples/toggle.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+use cortex_m_rt::entry;
+use stm32f3xx_hal::hal::digital::v2::OutputPin;
+use stm32f3xx_hal::prelude::*;
+use stm32f3xx_hal::stm32;
+
+#[entry]
+fn main() -> ! {
+    let p = stm32::Peripherals::take().unwrap();
```
Comment (dfrankland) [inline examples/toggle.rs]: To stay consistent with:
https://github.com/stm32-rs/stm32f3xx-hal/blob/902b56f8f11221cb8ffbf24305d66ecea858bf6c/examples/usb_serial.rs#L17

```suggestion
    let dp = stm32::Peripherals::take().unwrap();
```

File: examples/toggle.rs (hunk 1 above)
Comment (Sh3Rm4n) [reply examples/toggle.rs]: Oh yeah right :grinning: I've not considered, that the led colors are not consistent between the discovery boards.

File: examples/toggle.rs (hunk 2 above)
Comment (Sh3Rm4n) [reply examples/toggle.rs]: Personally, I am more in favor of the `use panic_semihosting as _;` method, as this is more Edition 2018 idiomatic (?)

However, consistency is more important, and I will happily change it.

File: src/prelude.rs (hunk 3 above)
Comment (Sh3Rm4n) [reply src/prelude.rs]: Yeah, you are right. I've considered it myself. Btw. is the cfg attribute always line based? Or is there another to apply it to a whole block / paragraph?

Comment (dfrankland) [reply src/prelude.rs]: So far as I know, it's line based only. I've used [`cfg_if`](https://github.com/alexcrichton/cfg-if) with success for block based application.

**Comment by Sh3Rm4n** [comment]
Applied change requests. Ready for review again :)

**Comment by dfrankland** [comment]
Nice work 🎉

**Pull request merged**